jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -127,52 +137,96 @@ def read_flag_byte(val):
             obj["ins"][i]["txinwitness"] = items
 
     obj["locktime"] = read_as_int(4)
     return obj
 
-def serialize(txobj):
-    o = []
-    if json_is_base(txobj, 16):
-        json_changedbase = json_changebase(txobj,
-                                           lambda x: binascii.unhexlify(x))
-        hexlified = safe_hexlify(serialize(json_changedbase))
-        return hexlified
-    o.append(encode(txobj["version"], 256, 4)[::-1])
+def serialize(tx):
+    """Assumes a deserialized transaction in which all
+    dictionary values are decoded hex strings or numbers.
+    Rationale: mixing raw bytes/hex/strings in dict objects causes
+    complexity; since the dict tx object has to be inspected
+    in this function, avoid complicated logic elsewhere by making
+    all conversions to raw byte strings happen here.
+    Table of dictionary keys and type for the value:
+    ================================================
+    version: int
+    ins: list
+    ins[0]["outpoint"]["hash"]: hex encoded string
+    ins[0]["outpoint"]["index"]: int
+    ins[0]["script"]: hex encoded string
+    ins[0]["sequence"]: int
+    ins[0]["txinwitness"]: list (optional, may not exist)
+    ins[0]["txinwitness"][0]: hex encoded string
+    outs: list
+    outs[0]["script"]: hex encoded string
+    outs[0]["value"]: int
+    locktime: int
+    =================================================
+    Returned serialized transaction is a byte string.
+    """
+    #Because we are manipulating the dict in-place, need
+    #to work on a copy
+    txobj = copy.deepcopy(tx)
+    o = BytesIO()
+    hexout = False
+    o.write(struct.pack(b'<i', txobj["version"]))
     segwit = False
-    if any("txinwitness" in x.keys() for x in txobj["ins"]):
+    if any("txinwitness" in x for x in txobj["ins"]):
         segwit = True
     if segwit:
         #append marker and flag
-        o.append('\x00')
-        o.append('\x01')
-    o.append(num_to_var_int(len(txobj["ins"])))
+        o.write(b'\x00')
+        o.write(b'\x01')
+    o.write(num_to_var_int(len(txobj["ins"])))
     for inp in txobj["ins"]:
-        o.append(inp["outpoint"]["hash"][::-1])
-        o.append(encode(inp["outpoint"]["index"], 256, 4)[::-1])
-        o.append(num_to_var_int(len(inp["script"])) + (inp["script"] if inp[
-            "script"] or is_python2 else bytes()))
-        o.append(encode(inp["sequence"], 256, 4)[::-1])
-    o.append(num_to_var_int(len(txobj["outs"])))
+        if len(inp["outpoint"]["hash"]) == 64:
+            o.write(binascii.unhexlify(inp["outpoint"]["hash"])[::-1])
+            hexout = True
+        elif len(inp["outpoint"]["hash"]) == 32:
+            o.write(inp["outpoint"]["hash"][::-1])
+        o.write(struct.pack(b'<I', inp["outpoint"]["index"]))
+        if len(inp["script"]) == 0:
+            o.write(b'\x00')
+        elif isinstance(inp["script"], basestring) and re.match('^[0-9a-fA-F]*$', inp["script"]):
```

Comment:
This function is already pretty hard to read and having this regex check everywhere doesn't really aid readability. Can this whole if/else be moved into a helper function?

Reply:
Ultimately I plan to remove these regex checks, I was trying to pull the type checks directly into the functions in order to isolate the conversion behavior as much as possible so that it can be incrementally refactored out, the main issue is that large parts of the codebase(especially in jmclient) expect automatic conversion behavior. Once py2 support is dropped I will be able to refactor this more easily.